## Notebook 1: Raw Ingestion

Goal: Load the raw Lending Club data and look at it, without cleaning or modifying anything yet.

Dataset: All Lending Club loan data (2007–2018 accepted & rejected loans)
https://www.kaggle.com/datasets/wordsforthewise/lending-club

We only use the accepted loans file here, since it's the only one with an outcome
(loan_status) — the rejected-loans file has no performance data.

### SETUP

Installing the Kaggle CLI and dotenv package for use with .env

In [ ]:
!pip install -q --upgrade kaggle python-dotenv

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.2/126.2 kB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 256.8/256.8 kB 10.4 MB/s eta 0:00:00


### AUNTHENTICATING Kaggle VIA `.env`


`.env` keeps our Kaggle credentials instaed of us harcoding them in the notebook. It should contain:
```
KAGGLE_USERNAME=your_kaggle_username
KAGGLE_KEY=your_kaggle_api_key
```

Colab doesn't have access to our local files by default, so we need to upload our `.env`. We do so via the prompt below.

In [ ]:
from google.colab import files

# Upload the .env file when prompted
uploaded = files.upload()

In [ ]:
from dotenv import load_dotenv
import os

load_dotenv()  # reads .env if file is named .env exactly, and loads KAGGLE_USERNAME / KAGGLE_KEY into the environment. If named for example kaggle.env, use load_dotenv('kaggle.env')

# quick check that the credentials were actually picked up
print("KAGGLE_USERNAME loaded:", "KAGGLE_USERNAME" in os.environ)
print("KAGGLE_KEY loaded:", "KAGGLE_KEY" in os.environ)

### DOWNLOADING THE DATASET and EXTRACTING

In [ ]:
!kaggle datasets download -d wordsforthewise/lending-club

Dataset URL: https://www.kaggle.com/datasets/wordsforthewise/lending-club
License(s): CC0-1.0
lending-club.zip: Skipping, found more recently modified local copy (use --force to force download)


In [ ]:
!unzip -n -q lending-club.zip

Archive:  lending-club.zip
  inflating: accepted_2007_to_2018Q4.csv.gz  
  inflating: accepted_2007_to_2018q4.csv/accepted_2007_to_2018Q4.csv  
  inflating: rejected_2007_to_2018Q4.csv.gz  
  inflating: rejected_2007_to_2018q4.csv/rejected_2007_to_2018Q4.csv  


In [ ]:
!ls -lh

total 1.9G
drwxr-xr-x 2 root root 4.0K Aug  1 09:13 accepted_2007_to_2018q4.csv
-rw-r--r-- 1 root root 375M Dec 17  2019 accepted_2007_to_2018Q4.csv.gz
-rw-r--r-- 1 root root 1.3G Dec 17  2019 lending-club.zip
drwxr-xr-x 2 root root 4.0K Aug  1 09:13 rejected_2007_to_2018q4.csv
-rw-r--r-- 1 root root 244M Dec 17  2019 rejected_2007_to_2018Q4.csv.gz
drwxr-xr-x 1 root root 4.0K Jun  4 13:32 sample_data


### IMPORTING NECESSARY LIBRARIES

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
# Remove the display limit for columns
pd.set_option('display.max_columns', None)

### LOADING DATASET

In [ ]:
SAMPLE_SIZE = int(input("Enter you sample size of choice in whole numbers: "))

df = pd.read_csv(
    "accepted_2007_to_2018Q4.csv.gz",
    compression="gzip",
    nrows=SAMPLE_SIZE,
    low_memory=False
)

Enter you sample size of choice in whole numbers: 150000


In [ ]:
# First 5 rows
df.head()

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,68407277,NaN,3600.0,3600.0,3600.0,36 months,13.99,123.03,C,C4,leadman,10+ years,MORTGAGE,55000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,debt_consolidation,Debt consolidation,190xx,PA,5.91,0.0,Aug-2003,675.0,679.0,1.0,30.0,NaN,7.0,0.0,2765.0,29.7,13.0,w,0.00,0.00,4421.723917,4421.72,3600.00,821.72,0.0,0.0,0.0,Jan-2019,122.67,NaN,Mar-2019,564.0,560.0,0.0,30.0,1.0,Individual,NaN,NaN,NaN,0.0,722.0,144904.0,2.0,2.0,0.0,1.0,21.0,4981.0,36.0,3.0,3.0,722.0,34.0,9300.0,3.0,1.0,4.0,4.0,20701.0,1506.0,37.2,0.0,0.0,148.0,128.0,3.0,3.0,1.0,4.0,69.0,4.0,69.0,2.0,2.0,4.0,2.0,5.0,3.0,4.0,9.0,4.0,7.0,0.0,0.0,0.0,3.0,76.9,0.0,0.0,0.0,178050.0,7746.0,2400.0,13734.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,68355089,NaN,24700.0,24700.0,24700.0,36 months,11.99,820.28,C,C1,Engineer,10+ years,MORTGAGE,65000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,small_business,Business,577xx,SD,16.06,1.0,Dec-1999,715.0,719.0,4.0,6.0,NaN,22.0,0.0,21470.0,19.2,38.0,w,0.00,0.00,25679.660000,25679.66,24700.00,979.66,0.0,0.0,0.0,Jun-2016,926.35,NaN,Mar-2019,699.0,695.0,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,204396.0,1.0,1.0,0.0,1.0,19.0,18005.0,73.0,2.0,3.0,6472.0,29.0,111800.0,0.0,0.0,6.0,4.0,9733.0,57830.0,27.1,0.0,0.0,113.0,192.0,2.0,2.0,4.0,2.0,NaN,0.0,6.0,0.0,5.0,5.0,13.0,17.0,6.0,20.0,27.0,5.0,22.0,0.0,0.0,0.0,2.0,97.4,7.7,0.0,0.0,314017.0,39475.0,79300.0,24667.0,NaN,NaN,NaN,NaN,NaN,N

In [ ]:
# checking the shape of the dataset
df.shape

(150000, 151)

In [ ]:
# detailed info; name, non-empty values, data type
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 151 columns):
 #    Column                                      Non-Null Count   Dtype  
---   ------                                      --------------   -----  
 0    id                                          150000 non-null  int64  
 1    member_id                                   0 non-null       float64
 2    loan_amnt                                   150000 non-null  float64
 3    funded_amnt                                 150000 non-null  float64
 4    funded_amnt_inv                             150000 non-null  float64
 5    term                                        150000 non-null  object 
 6    int_rate                                    150000 non-null  float64
 7    installment                                 150000 non-null  float64
 8    grade                                       150000 non-null  object 
 9    sub_grade                                   150000 non-nu

In [ ]:
print(f"Total columns: {df.shape[1]}\n")
for index, col in enumerate(df.columns):
  print(f"{index}: {col}")

Total columns: 151

0: id
1: member_id
2: loan_amnt
3: funded_amnt
4: funded_amnt_inv
5: term
6: int_rate
7: installment
8: grade
9: sub_grade
10: emp_title
11: emp_length
12: home_ownership
13: annual_inc
14: verification_status
15: issue_d
16: loan_status
17: pymnt_plan
18: url
19: desc
20: purpose
21: title
22: zip_code
23: addr_state
24: dti
25: delinq_2yrs
26: earliest_cr_line
27: fico_range_low
28: fico_range_high
29: inq_last_6mths
30: mths_since_last_delinq
31: mths_since_last_record
32: open_acc
33: pub_rec
34: revol_bal
35: revol_util
36: total_acc
37: initial_list_status
38: out_prncp
39: out_prncp_inv
40: total_pymnt
41: total_pymnt_inv
42: total_rec_prncp
43: total_rec_int
44: total_rec_late_fee
45: recoveries
46: collection_recovery_fee
47: last_pymnt_d
48: last_pymnt_amnt
49: next_pymnt_d
50: last_credit_pull_d
51: last_fico_range_high
52: last_fico_range_low
53: collections_12_mths_ex_med
54: mths_since_last_major_derog
55: policy_code
56: application_type
57: annual_in

### MISSING VALUES

In [ ]:
missing_values = df.isna().sum().sort_values(ascending=False) # missing count
missing_values_pct = (missing_values / len(df)*100).round(2)    # missing percentage per count
missing_summary = pd.DataFrame({"Missing Count": missing_values, "Missing Percentage": missing_values_pct})
missing_summary[missing_summary["Missing Count"] > 0]           # only columns with missing data

,Missing Count,Missing Percentage
member_id,150000,100.00
sec_app_num_rev_accts,150000,100.00
sec_app_open_act_il,150000,100.00
sec_app_inq_last_6mths,150000,100.00
sec_app_open_acc,150000,100.00
sec_app_mort_acc,150000,100.00
sec_app_mths_since_last_major_derog,150000,100.00
sec_app_collections_12_mths_ex_med,150000,100.00
sec_app_chargeoff_within_12_mths,150000,100.00
sec_app_fico_range_low,150000,100.00


In [ ]:
# Columns that are mostly empty (>90% missing) — likely to drop later
mostly_empty = missing_summary[missing_summary["Missing Percentage"] > 90]
print(f"{len(mostly_empty)} columns are more than 90% missing:")
mostly_empty

38 columns are more than 90% missing:


,Missing Count,Missing Percentage
member_id,150000,100.00
sec_app_num_rev_accts,150000,100.00
sec_app_open_act_il,150000,100.00
sec_app_inq_last_6mths,150000,100.00
sec_app_open_acc,150000,100.00
sec_app_mort_acc,150000,100.00
sec_app_mths_since_last_major_derog,150000,100.00
sec_app_collections_12_mths_ex_med,150000,100.00
sec_app_chargeoff_within_12_mths,150000,100.00
sec_app_fico_range_low,150000,100.00


### TARGET VARIABLE

In [ ]:
# how many times each unique value appears in our target variable
df["loan_status"].value_counts(dropna=False)

,count
loan_status,
Fully Paid,105691
Charged Off,26330
Current,16982
Late (31-120 days),611
In Grace Period,271
Late (16-30 days),114
Default,1


In [ ]:
# What proportion of the 100%?
df['loan_status'].value_counts(normalize=True, dropna=False).round(4) * 100

,proportion
loan_status,
Fully Paid,70.46
Charged Off,17.55
Current,11.32
Late (31-120 days),0.41
In Grace Period,0.18
Late (16-30 days),0.08
Default,0.00


### USEFUL COLUMNS that relate to loan performance

In [ ]:
shortlisted_features = [
    'loan_amnt', 'term', 'int_rate', 'installment', 'grade', 'sub_grade',
    'emp_length', 'home_ownership', 'annual_inc', 'verification_status',
    'issue_d', 'loan_status', 'purpose', 'addr_state', 'dti',
    'delinq_2yrs', 'fico_range_low', 'fico_range_high', 'open_acc',
    'pub_rec', 'revol_bal', 'revol_util', 'total_acc'
]

# safety check
missing_from_df = []

for feature in shortlisted_features:
  if feature not in df.columns:
    missing_from_df.append(feature)

if missing_from_df:
  print(f"Shorlisted Columns not in the DataFrame: {missing_from_df}")
else:
  print("All shortlisted columns are present in the DataFrame.")

All shortlisted columns are present in the DataFrame.


### SUMMARY STATS

In [ ]:
# DataFrame with shortlisted features
df_shortlisted = df[shortlisted_features]

df_shortlisted.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
loan_amnt,150000.0,NaN,NaN,NaN,15189.246,8647.245331,1000.0,8400.0,14000.0,20000.0,35000.0
term,150000,2,36 months,101503,NaN,NaN,NaN,NaN,NaN,NaN,NaN
int_rate,150000.0,NaN,NaN,NaN,12.282534,4.204552,5.32,9.17,12.05,14.65,28.99
installment,150000.0,NaN,NaN,NaN,438.627423,246.685408,14.77,259.36,382.55,577.44,1445.46
grade,150000,7,B,45134,NaN,NaN,NaN,NaN,NaN,NaN,NaN
sub_grade,150000,35,C1,9651,NaN,NaN,NaN,NaN,NaN,NaN,NaN
emp_length,140965,11,10+ years,50764,NaN,NaN,NaN,NaN,NaN,NaN,NaN
home_ownership,150000,4,MORTGAGE,74675,NaN,NaN,NaN,NaN,NaN,NaN,NaN
annual_inc,150000.0,NaN,NaN,NaN,78706.129924,86845.384346,0.0,48000.0,65641.0,94000.0,9000000.0
verification_status,150000,3,Source Verified,63788,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### CHECKPOINT for NoteBook 2

In [ ]:
RAW_OUT_DIR = '../data/raw'     # folder 'data' and subfolder 'raw'
os.makedirs(RAW_OUT_DIR, exist_ok=True)   # create folder if non-existent, else continue with no error
raw_out_path = f'{RAW_OUT_DIR}/accepted_loans_sample_raw.parquet'   # name created folder
df.to_parquet(raw_out_path, index=False)      # save data, where parquet is better at saving tabular data than csv files

print(f"Saved raw sample: {df.shape[0]:,} rows x {df.shape[1]} columns -> {raw_out_path}")    # confirm the save

Saved raw sample: 150,000 rows x 151 columns -> ../data/raw/accepted_loans_sample_raw.parquet


## What I Observed


> Total columns / rows in the sample:
- 150,000 rows × 151 columns



> Columns that are mostly empty (candidates to drop):
- Out of the `151` columns that are in the dataset, `73` of them have values missing. Out of these 73 columns, there are `38` columns that are more than `90%` empty.
- Upon investigation, I found out that most of these empty columns belonged to these `three` groups:
   1. **Second applicant information:** sec_app_*, revol_bal_joint, dti_joint, etc.
   2. **Hardship information:** hardship_type, hardship_reason, hardship_status, etc.
   3. **Debt settlement information:** settlement_status, settlement_amount, etc.
  
- The verdict was that the Second applicant information columns were mostly empty because very few loans in the dataset have a second applicant, the Hardship information ones because only a small number of borrowers needed or joined a hardship program, and the Debt settlement information ones because only a small number of loans went through debt settlement.
- I decided to drop these 38 columns in Notebook 2, not because the data is bad or incorrect, but simply because they are not useful for our analysis. They contain information for only a very small number of loans, and keeping them would add unnecessary columns without giving us much useful information.



> How `loan_status` breaks down, and how I'll collapse it into Good/Bad in notebook 2:
- The loan_status column tells us what happened to each loan, and here is the breakdown:
    1. **Fully Paid**: `105,691 loans` (70.46%) → the loan was successfully paid back.
    2. **Charged Off**: `26,330 loans` (17.55%) → the loan was not fully repaid.
    3. **Current**: `16,982 loans` (11.32%) → the loan is still ongoing.
    4. **Late (31–120 days**): `611 loans` (0.41%) → the borrower is seriously late with payments.
    5. **In Grace Period**: `271 loans` (0.18%) → the loan is temporarily within a payment grace period.
    6. **Late (16–30 days)**: `114 loans` (0.08%) → the borrower is late with payments.
    7. **Default**: `1 loan` (0.00%) → the loan has officially defaulted.

- To make the analysis easier, I'll group the loans into three simple categories:
    1. **Good:**
        - Fully Paid

    2. **Bad:** (all loans that failed to be repaid on schedule)
       - Charged Off
       - Default
       - Late (31–120 days)
       - Late (16–30 days)

    3. **Still Active**
       - Current
       - In Grace Period

- I will be keeping **Current** and **In Grace Period** separate because these loans are not finished yet, and we don't know whether they will eventually be paid successfully or become bad loans. Putting them into either **Good** or **Bad** would mean that we are already guessing the final outcome.
- For the main analysis in Notebook 3, I'll therefore focus on loans with a clear outcome: **Good vs. Bad**, while leaving the **Still Active** loans out of that comparison.



> Data leakage risk:
- Some columns tell us what happened after the borrower received the loan.
   1. **total_pymnt** → how much the borrower eventually paid
   2. **total_rec_prncp** → how much of the original loan was repaid
   3. **total_rec_int** → how much interest was paid
   4. **recoveries** → money recovered after a loan went bad
   5. **last_pymnt_d** → date of the last payment
   6. **last_pymnt_amnt** → amount of the last payment
   7. **out_prncp** → amount of the loan that was still outstanding

- I am planning on not using these columns in the main analysis because they are the results of the loan, and not information we had when the loan was first given.
- **For example**, if we are trying to understand why a loan became `Bad`, using `total_pymnt` would not be fair because we already know how much the borrower paid. That information comes after the loan's performance has happened, and using it is like trying to predict whether a student will pass an exam by using their final exam score, but the answer is already revealed.
- These columns will be removed from the performance analysis, to make sure our analysis focuses on the original characteristics of the borrower and loan, rather than information that became available afterward. I'll only use them when needed to understand or confirm the final loan outcome.

